pilot/pkg/networking/core/v1alpha3/listener_builder.go

```diff
@@ -37,11 +40,76 @@ type ListenerBuilder struct {
 	managementListeners    []*xdsapi.Listener
 	virtualListener        *xdsapi.Listener
 	virtualInboundListener *xdsapi.Listener
+	// true inbound listener use filter chain match
+	useInboundFilterChain bool
+	// Unimplemented
+	useOutboundFilterChain bool
+}
+
+func reduceInboundListenerToFilters(listeners []*xdsapi.Listener) (chains []*listener.FilterChain, needTLS bool) {
+	needTLS = false
+	chains = make([]*listener.FilterChain, 0)
+	for _, l := range listeners {
+		for _, c := range l.FilterChains {
+			chain := c
+			mergeFilterChainFromInboundListener(&chain, l, &needTLS)
+			chains = append(chains, &chain)
+		}
+	}
+	// TODO(silentdai): sort
+	return
+}
+
+func (builder *ListenerBuilder) aggregateVirtualInboundListener(env *model.Environment, node *model.Proxy) *ListenerBuilder {
+	var isTransparentProxy *google_protobuf.BoolValue
+	if node.GetInterceptionMode() == model.InterceptionTproxy {
+		isTransparentProxy = proto.BoolTrue
+	}
+	tcpProxyFilter := newTCPProxyListenerFilter(env, node, true)
+	filterChains, needTLS := reduceInboundListenerToFilters(builder.inboundListeners)
+	actualWildcard, _ := getActualWildcardAndLocalHost(node)
+
+	builder.virtualInboundListener = &xdsapi.Listener{
+		Name:        VirtualInboundListenerName,
+		Address:     util.BuildAddress(actualWildcard, ProxyInboundListenPort),
+		Transparent: isTransparentProxy,
+		ListenerFilters: []listener.ListenerFilter{
+			{
+				Name: "envoy.listener.original_dst",
+			},
+		},
+		// Deprecated by envoyproxy. Replaced
+		// 1. filter chains in this listener
+		// 2. explicit original_dst listener filter
+		// UseOriginalDst: proto.BoolTrue,
+		// TODO(silentdai) remove the fallback tcp filter chain and integrate sniff filter chain
+		FilterChains: []listener.FilterChain{
+			{
+				Filters: []listener.Filter{*tcpProxyFilter},
+			},
+		},
+	}
+
+	for _, c := range filterChains {
+		builder.virtualInboundListener.FilterChains =
+			append(builder.virtualInboundListener.FilterChains, *c)
+	}
+	if needTLS {
+		builder.virtualInboundListener.ListenerFilters =
+			append(builder.virtualInboundListener.ListenerFilters, listener.ListenerFilter{
+				Name: "envoy.listener.tls_inspector",
+			})
+	}
+	return builder
 }
 
 func NewListenerBuilder(node *model.Proxy) *ListenerBuilder {
 	builder := &ListenerBuilder{
 		node: node,
+		// The extra inbound listener has no side effect for iptables that doesn't redirect to 15006
+		useInboundFilterChain: true,
```

Comment:
Then why do we add it ?

Reply:
The idea is we can actually create 15006 but disable it in iptables :)